hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/ServiceFactories.java

```diff
@@ -98,6 +98,17 @@ private ServiceFactories() { }
      * A variant of {@link #sharedService(FunctionEx, ConsumerEx)
      * sharedService(createFn, destroyFn)} with a no-op {@code
      * destroyFn}.
+     * <p>
+     * <strong>Note:</strong> if your service has a blocking API (e.g., doing
+     * synchronous IO or acquiring locks), you must call {@link
+     * ServiceFactory#toNonCooperative()} as a hint to the Jet execution engine
+     * to start a dedicated thread for those calls. Failing to do this can
+     * cause severe performance problems. You should also carefully consider
+     * how much local parallelism you need for this step since each parallel
+     * tasklet needs its own thread. Call {@code GeneralStage#setLocalParallelism
+     * stage.setLocalParallelism()} to set an explicit level, otherwise it will
+     * depend on the number of cores on the Jet machine, which makes no sense
+     * for blocking code.
```

Comment:
I think it would be good to point to `Processor.isCooperative()` javadoc. The _cooperative contract_ is explained there in more detail.

Reply:
`ServiceFactory#toNonCooperative` leads to there already.